lib/Sema/CSBindings.cpp

```diff
@@ -366,11 +366,59 @@ void PotentialBindings::inferTransitiveBindings(
   }
 }
 
+// If potential binding or any of it's adjacents variable bindings is a closure
+// that has a subtype relation associated with argument conversion constraint
+// located directly on an autoclosure parameter.
+static bool isClosureInAutoClosureArgumentConversion(
+    PotentialBindings &currentBindings,
+    llvm::SmallDenseMap<TypeVariableType *, PotentialBindings>
+        &inferredBindings) {
+  auto isClosureVarInAutoClosureContext = [](PotentialBindings &bindings) {
+    if (!bindings.TypeVar->getImpl().isClosureType())
+      return false;
+
+    return llvm::any_of(
+        bindings.SubtypeOf,
+        [](std::pair<TypeVariableType *, Constraint *> subType) {
+          if (subType.second->getKind() != ConstraintKind::ArgumentConversion)
+            return false;
+          return subType.second->getLocator()
+              ->isLastElement<LocatorPathElt::AutoclosureResult>();
+        });
+  };
+
+  if (isClosureVarInAutoClosureContext(currentBindings))
+    return true;
+
+  return llvm::any_of(
```

Comment:
Looking at this I wonder whether we need to get DefaultClosureType transitively inferred at all, I think it should be limited to `Defaultable` only...

Reply:
Hum, so the transitive type variable should known the subtypes bindings as well? So it would be possible to just check the binding `Subtypes`?

Reply:
By the test I run, it we don't check for the adjacents the default is still inferred for the type var ... because at the point of checking `SubtypeOf` for the var that is transitively inferred was empty. Is that what you mean? Sorry if I'm misunderstanding ...

Reply:
Or do you mean do we check only for the type var and remove the defaults from all `inferredBindings`?

Reply:
I mean that `inferTransitiveBindings` shouldn't copy any defaults besides `Defaultable` up the subtype chain, that way you wouldn't need to check adjacencies.

Reply:
Ah, I see... but that wouldn't affect any other scenario that may rely on this behavior? I just worried if it would be safe to do that

Reply:
Conceptually nothing besides the closure type variable should use that default type anyway, so what is happening today might be incorrect.

Reply:
Ah ok, I just made the changes :)

Reply:
Great! Let's run source compat suite and regular tests.

Reply:
Awesome!